### PR TITLE
Add --strict option to build

### LINKED ISSUE #1334: Allow treating warnings as errors during packaging / validation
This will allow for a 'no warnings' policy in an automated release pipeline, and would help prevent releasing broken gems

### PR DESCRIPTION
**Pull request by segiddins**
Closes #1334

---

This change is [Reviewable](https://reviewable.io/reviews/rubygems/rubygems/1335)

### CODE REVIEW

File: lib/rubygems/specification.rb [hunk 1]
```diff
@@ -2822,7 +2822,7 @@ def validate_dependencies # :nodoc:
       end
 
       warning_messages << "prerelease dependency on #{dep} is not recommended" if
-        prerelease_dep
+        prerelease_dep && !version.prerelease?
```
Comment (drbrain) [inline lib/rubygems/specification.rb]: Please move this to a separate pull request as it is separate from disallowing warnings.

File: lib/rubygems/package.rb [hunk 2]
```diff
@@ -239,12 +239,14 @@ def add_metadata tar # :nodoc:
...
 
-  def build skip_validation = false
+  def build skip_validation = false, allow_warnings = true
     Gem.load_yaml
     require 'rubygems/security'
 
     @spec.mark_version
-    @spec.validate unless skip_validation
+    unless skip_validation || !@spec.validate(true, allow_warnings)
+      raise Gem::InvalidSpecificationException, "Spec failed to validate"
```
Comment (drbrain) [inline lib/rubygems/package.rb]: Raising here means `allow_warnings` is a no-op for users who `spec.validate anything, true`.  Raising this exception should move into Gem::Specification#validate.

File: lib/rubygems/package.rb [hunk 3]
```diff
@@ -239,12 +239,14 @@ def add_metadata tar # :nodoc:
   ##
   # Builds this package based on the specification set by #spec=
 
-  def build skip_validation = false
+  def build skip_validation = false, allow_warnings = true
```
Comment (drbrain) [inline lib/rubygems/package.rb]: `package.build spec, true, false` is not a valid combination of arguments, checking it here seems best.

File: lib/rubygems/commands/build_command.rb [hunk 4]
```diff
@@ -9,6 +9,10 @@ def initialize
     add_option '--force', 'skip validation of the spec' do |value, options|
       options[:force] = true
     end
+
+    add_option '--[no-]allow-warnings', 'allow warnings when validating the spec' do |value, options|
```
Comment (drbrain) [inline lib/rubygems/commands/build_command.rb]: Here are some examples of how to test this option: https://github.com/rubygems/rubygems/blob/d3db595be39639fdc6b020e24fab4d0cf052b448/test/rubygems/test_gem_commands_check_command.rb#L29-L45

File: lib/rubygems/commands/build_command.rb [hunk 5]
```diff
@@ -45,7 +49,7 @@ def execute
       spec = Gem::Specification.load gemspec
 
       if spec then
-        Gem::Package.build spec, options[:force]
+        Gem::Package.build spec, options[:force], options[:allow_warnings]
```
Comment (drbrain) [inline lib/rubygems/commands/build_command.rb]: You should be able to use [this test](https://github.com/rubygems/rubygems/blob/master/test/rubygems/test_gem_commands_build_command.rb#L27-L51) as a base for testing `--allow-warnings`

File: lib/rubygems/package.rb (hunk 3 above)
Comment (drbrain) [inline lib/rubygems/package.rb]: You can use [this test](https://github.com/rubygems/rubygems/blob/master/test/rubygems/test_gem_package.rb#L282-L293) as a base for testing `package.build false, true`

File: lib/rubygems/specification.rb [hunk 6]
```diff
@@ -2789,7 +2789,7 @@ def validate packaging = true
 
     validate_dependencies
 
-    true
+    allow_warnings || @warnings.zero?
```
Comment (drbrain) [inline lib/rubygems/specification.rb]: You can use [this test](https://github.com/rubygems/rubygems/blob/master/test/rubygems/test_gem_specification.rb#L2733-L2741) as a base for testing `spec.validate true, false`. (#validate should raise.)

**Comment by drbrain** [comment]
I think I would prefer changing [Gem::Specification#warning](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/specification.rb#L2951-L2955) to raise the exception if `allow_warnings` is false, but that's a slightly more complicated change to Gem::Specification, including changing the name of the `warning` method (perhaps to `validation_warning`).

File: lib/rubygems/package.rb (hunk 2 above)
Comment (segiddins) [inline lib/rubygems/package.rb]: It's not a no-op -- it changes the return value of `validate` to be `false` if there are warnings

Comment (drbrain) [inline lib/rubygems/package.rb]: As a user of the `allow_warnings` option for Gem::Specification#validate I expect it to raise an exception when I set the flag for turning warnings into errors.

With your patch there are three results for an API user to handle:

1. No validation error, you can package the gem (but possible warning)
2. Validation error, you cannot package the gem, use `rescue` to inform their user.
3. Warning detected as an error, you cannot package the gem, you must now check the return value to know this.

If #validate returns `false` when an API user calls it directly (not through Gem::Package.build) with `allow_warnings` set to false then they have to reimplement raising an error or special case their own code in order to display the error correctly.

In order to discover this behavior they must read carefully, which is something users don't do.

By making #validate uniform in its handling of invalid portions of the specification users no longer have to read which reduces maintenance burden and makes use of the option's implementation and use much cleaner.  For example, we no longer need expressions with double-negatives like:

`unless skip_validation || !@spec.validate(true, allow_warnings)`

**Comment by drbrain** [comment]
After thinking, I find `--allow-warnings` and `allow_warnings = true` unclear.  I'm uncertain if it means "warnings are errors" or "ignore warnings".

I think `--warnings-are-errors` and `warnings_are_errors` are clearer names to use since they describe the behavior accurately.

**Comment by djberg96** [comment]
I like --strict. :)

**Comment by indirect** [comment]
@segiddins this fails on every ruby, could you rebase/investigate? I also like `--strict` to mean warnings are errors.

**Comment by bronzdoc** [comment]
closing this in favor of https://github.com/rubygems/rubygems/pull/2332#issuecomment-400154670
